aws/resource_aws_sns_application.go

```diff
@@ -0,0 +1,243 @@
+
+	attributes["PlatformCredential"] = aws.String(d.Get("credential").(string))
+
+	if _, ok := SupportedPlatforms[platform]; !ok {
+		return errors.New(fmt.Sprintf("Platform %s is not supported", platform))
+	}
+
+	if value, _ := SupportedPlatforms[platform]; value {
+		if principal == "" {
+			return errors.New(fmt.Sprintf("Principal is required for %s", platform))
+		} else {
+			attributes["PlatformPrincipal"] = aws.String(principal)
+		}
+	}
+
+	log.Printf("[DEBUG] SNS create application: %s", name)
+
+	req := &sns.CreatePlatformApplicationInput{
+		Name:       aws.String(name),
+		Platform:   aws.String(platform),
+		Attributes: attributes,
+	}
+
+	output, err := snsconn.CreatePlatformApplication(req)
+	if err != nil {
+		return fmt.Errorf("Error creating SNS application: %s", err)
+	}
+
+	d.SetId(*output.PlatformApplicationArn)
+
+	// Write the ARN to the 'arn' field for export
+	d.Set("arn", *output.PlatformApplicationArn)
+
+	return resourceAwsSnsApplicationUpdate(d, meta)
+}
+
+func resourceAwsSnsApplicationUpdate(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	resource := *resourceAwsSnsApplication()
+
+	attributes := make(map[string]*string)
+
+	for k, _ := range resource.Schema {
+		if attrKey, ok := SNSPlatformAppAttributeMap[k]; ok {
+			if d.HasChange(k) {
+				log.Printf("[DEBUG] Updating %s", attrKey)
+				_, n := d.GetChange(k)
+				attributes[attrKey] = aws.String(n.(string))
+			}
+		}
+	}
+
+	if d.HasChange("credential") {
+		attributes["PlatformCredential"] = aws.String(d.Get("credential").(string))
+		// If the platform requires a principal it must also be specified, even if it didn't change
+		// since credential is stored as a hash, the only way to update principal is to update both
+		// as they must be specified together in the request.
+		if v, _ := SupportedPlatforms[d.Get("platform").(string)]; v {
+			attributes["PlatformPrincipal"] = aws.String(d.Get("principal").(string))
+		}
+	}
+
+	// Make API call to update attributes
+	req := &sns.SetPlatformApplicationAttributesInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+		Attributes:             attributes,
+	}
+	_, err := snsconn.SetPlatformApplicationAttributes(req)
+
+	if err != nil {
+		return fmt.Errorf("Error updating SNS application: %s", err)
+	}
+
+	return resourceAwsSnsApplicationRead(d, meta)
+}
+
+func resourceAwsSnsApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	attributeOutput, err := snsconn.GetPlatformApplicationAttributes(&sns.GetPlatformApplicationAttributesInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if attributeOutput.Attributes != nil && len(attributeOutput.Attributes) > 0 {
+		attrmap := attributeOutput.Attributes
+		resource := *resourceAwsSnsApplication()
+		// iKey = internal struct key, oKey = AWS Attribute Map key
+		for iKey, oKey := range SNSPlatformAppAttributeMap {
```

Comment:
This is a nice way of iterating through attributes that we can `Update`, but during the resources Read, we should be more verbose about each item that we read from the API request. 
Usually this means, _more_ code, but it's usually a lot clearer which attributes we're reading from the resource. This may require a separate request to populate the attributes that we're currently not setting. 
IE: Can a user currently update the resources `"name"`, `"platform"`, or `"credential"` in the AWS Console? (I'm assuming these cannot be updated, as they're currently listed as `ForceNew`).